cmd/sgx_plugin/sgx_plugin.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2020 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"path"
+	"runtime"
+	"strconv"
+
+	dpapi "github.com/intel/intel-device-plugins-for-kubernetes/pkg/deviceplugin"
+	"k8s.io/klog"
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+const (
+	// Device plugin settings.
+	namespace              = "sgx.intel.com"
+	deviceTypeEnclave      = "enclave"
+	deviceTypeProvision    = "provision"
+	devicePath             = "/dev"
+	podsPerCoreEnvVariable = "PODS_PER_CORE"
+	defaultPodsPerCore     = 10
+)
+
+type devicePlugin struct {
+	devfsDir   string
+	nEnclave   uint
+	nProvision uint
+	scanDone   chan bool
+}
+
+func newDevicePlugin(devfsDir string, nEnclave, nProvision uint) *devicePlugin {
+	return &devicePlugin{
+		devfsDir:   devfsDir,
+		nEnclave:   nEnclave,
+		nProvision: nProvision,
+		scanDone:   make(chan bool, 1),
+	}
+}
+
+func (dp *devicePlugin) Scan(notifier dpapi.Notifier) error {
+	devTree, err := dp.scan()
+	if err != nil {
+		return err
+	}
+	notifier.Notify(devTree)
+
+	// Wait forever to prevent manager run loop from exiting.
+	<-dp.scanDone
```

Comment:
Is it going to be one-time scan always? I'm thinking of the case for loading/unloading the kernel driver.

Reply:
I believe that SGX driver can't be built as module, so it shouldn't be possible to load or unload.

Reply:
correct, it can't be built as a module (`CONFIG_INTEL_SGX` is `bool`)